src/Microsoft.AspNetCore.Blazor.Browser.JS/src/Interop/InvokeDotNetMethodWithJsonMarshalling.ts

```diff
@@ -0,0 +1,183 @@
+import { platform } from '../Environment';
+import { System_String, Pointer } from '../Platform/Platform';
+import { getRegisteredFunction } from './RegisteredFunction';
+import { error } from 'util';
+
+export interface MethodOptions {
+  type: TypeInstance;
+  method: MethodInstance;
+}
+
+// Keep in sync with InvocationResult.cs
+export interface InvocationResult {
+  succeeded: boolean;
+  result?: any;
+  message?: string;
+}
+
+export interface MethodInstance {
+  name: string;
+  typeArguments?: { [key: string]: TypeInstance }
+  parameterTypes?: TypeInstance[];
+}
+
+export interface TypeInstance {
+  assembly: string;
+  name: string;
+  typeArguments?: { [key: string]: TypeInstance };
+}
+
+export function invokeDotNetMethod<T>(methodOptions: MethodOptions, ...args: any[]): (T | null) {
+  return invokeDotNetMethodCore(methodOptions, null, ...args);
+}
+
+let registrations = {};
+
+function resolveRegistration(methodOptions: MethodOptions) {
+  let existing = registrations[methodOptions.type.assembly];
+  existing = existing && registrations[methodOptions.type.name];
+  existing = existing && registrations[methodOptions.method.name];
+  if (existing !== undefined) {
+    return existing;
+  } else {
+    const method = platform.findMethod(
```

Comment:
Minor point, but can we cache this method lookup?

Reply:
Can we? It's a method_handle, I would imagine any handle can be moved by the GC

Reply:
The "method handle" is some identifier returned by the platform. It's not a pointer to a .NET object. So it's not subject to being moved by the GC compacter.

Or if it *is* a .NET object pointer, then we have a more widespread bug because the pattern of caching `findMethod` return values is used in a lot of places.

So yes please, let's cache the `findMethod` return values like elsewhere. If it turns out we have to change that globally we'll do so.